### ISSUE TITLE
`Layout/ClosingHeredocIndentation` & `Style/RedundantHeredocDelimiterQuotes` - undefined method `captures' for nil:NilClass

### ISSUE THREAD

**Issue by snutij** [rubocop/rubocop#11998]
Hello 👋 I open an issue for both cops, because it failed on the same method `heredoc_type` do not hesitate to tell me if you prefer two distinct issues.  

Given a file `foo.rb`: 
```ruby
<<""

```

Which is a valid syntax: 

```
ruby -c foo.rb
Syntax OK
```

--------

## Expected behavior

Should not trigger an error. I think this is a valid `heredoc` syntax, as it start with `<<` maybe some related offenses can be displayed. I'm very terrible with regex, so I can't dig more on group expectation etc. 😞  

## Actual behavior

`bin/rubocop -d --only Style/RedundantHeredocDelimiterQuotes foo.rb`
 
```
An error occurred while Style/RedundantHeredocDelimiterQuotes cop was inspecting foo.rb:1:0.
undefined method `captures' for nil:NilClass
/Users/snutij/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.53.1/lib/rubocop/cop/mixin/heredoc.rb:34:in `heredoc_type'
/Users/snutij/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.53.1/lib/rubocop/cop/style/redundant_heredoc_delimiter_quotes.rb:49:in `need_heredoc_delimiter_quotes?'
/Users/snutij/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.53.1/lib/rubocop/cop/style/redundant_heredoc_delimiter_quotes.rb:37:in `on_heredoc'
/Users/snutij/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.53.1/lib/rubocop/cop/mixin/heredoc.rb:12:in `on_str'
/Users/snutij/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.53.1/lib/rubocop/cop/commissioner.rb:107:in `public_send'
```

`bin/rubocop -d --only Style/ClosingHeredocIndentation foo.rb`

```
An error occurred while Layout/ClosingHeredocIndentation cop was inspecting foo.rb:1:0.
undefined method `captures' for nil:NilClass
/Users/snutij/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.53.1/lib/rubocop/cop/mixin/heredoc.rb:34:in `heredoc_type'
/Users/snutij/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.53.1/lib/rubocop/cop/layout/closing_heredoc_indentation.rb:59:in `on_heredoc'
/Users/snutij/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.53.1/lib/rubocop/cop/mixin/heredoc.rb:12:in `on_str'
/Users/snutij/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rubocop-1.53.1/lib/rubocop/cop/commissioner.rb:107:in `public_send'
```

## Steps to reproduce the problem

Both `Layout/ClosingHeredocIndentation` & `Style/RedundantHeredocDelimiterQuotes` are using the `heredoc_type` method from the module `RuboCop::Cop::Heredoc`

Which receive a node: 

```ruby
(rdbg) node
s(:dstr)
(ruby) node.source
"<<\"\""
```

Nothing match with the pattern: 
```OPENING_DELIMITER = /(<<[~-]?)['"`]?([^'"`]+)['"`]?/.freeze```

```ruby
(ruby) node.source.match(OPENING_DELIMITER)
nil
```

So the capture group `node.source.match(OPENING_DELIMITER).captures[0]` trigger an error.

https://github.com/rubocop/rubocop/blob/aaf1c3f6285e5e9e84d43c37944a7faaacb3bf60/lib/rubocop/cop/layout/closing_heredoc_indentation.rb#L57-L66

https://github.com/rubocop/rubocop/blob/aaf1c3f6285e5e9e84d43c37944a7faaacb3bf60/lib/rubocop/cop/style/redundant_heredoc_delimiter_quotes.rb#L48-L54

https://github.com/rubocop/rubocop/blob/aaf1c3f6285e5e9e84d43c37944a7faaacb3bf60/lib/rubocop/cop/mixin/heredoc.rb#L32-L34

## RuboCop version

```
bundle exec rubocop -V
1.53.1 (using Parser 3.2.2.3, rubocop-ast 1.29.0, running on ruby 3.2.2) [arm64-darwin22]
  - rubocop-minitest 0.31.0
  - rubocop-rake 0.6.0
  - rubocop-sorbet 0.7.0
```

**Comment by AlexWayfer** [comment]
This can be relevant: https://github.com/rubocop/rubocop/issues/12013
